pysmt/typing.py

```diff
@@ -280,6 +284,91 @@ def __hash__(self):
         return self._hash
 
 
+# ArrayType is a Factory that returns a _ArrayType
+def ArrayType(index_type, elem_type):
+    """Returns the singleton of the Array type with the given arguments.
+
+    This function takes care of building and registering the type
+    whenever needed. To see the functions provided by the type look at
+    _ArrayType
+    """
+    key = (index_type, elem_type)
+    if key in __ARRAY_TYPES__:
+        return  __ARRAY_TYPES__[key]
+
+    res = _ArrayType(index_type, elem_type)
+    __ARRAY_TYPES__[key] = res
+    return res
+
+
+class _ArrayType(PySMTType):
+    """Internal class used to represent an Array type.
+
+    This class should not be instantiated directly, but the factory
+    method ArrayType should be used instead.
+    """
+    def __init__(self, index_type, elem_type):
+        PySMTType.__init__(self, type_id = 5)
+        self._index_type = index_type
+        self._elem_type = elem_type
+        self._hash = hash(str(self))
+        return
+
+    @property
+    def elem_type(self):
+        """Returns the element type.
+
+        E.g.,  A: (Array Int Real)
+        Returns RealType.
+        """
+        return self._elem_type
+
+    @property
+    def index_type(self):
+        """Returns the index type.
+
+        E.g.,  A: (Array Int Real)
+        Returns IntType.
+        """
+        return self._index_type
+
+    def as_smtlib(self, funstyle=True):
+        itype = self.index_type.as_smtlib(False)
+        etype = self.elem_type.as_smtlib(False)
+
+        if funstyle:
+            return "() (Array %s %s)" % (itype, etype)
+        else:
+            return "(Array %s %s)" % (itype, etype)
+
+    def __str__(self):
+        return "ARR[%s -> %s]" % (self.index_type, self.elem_type)
+
+    def is_array_type(self):
+        return True
+
+    def __eq__(self, other):
+        if other is None:
+            return False
+        if self.type_id != other.type_id:
+            return False
+        if id(self) == id(other):
+            return True
+        return str(self) == str(other)
```

Comment:
This could be handled by recursive application of equality on the index and value types